zipkin/src/main/java/zipkin/storage/InMemorySpanStore.java

```diff
@@ -40,12 +38,33 @@
 
 /** Internally, spans are indexed on 64-bit trace ID */
 public final class InMemorySpanStore implements SpanStore {
-  private final Multimap<Long, Span> traceIdToSpans = new LinkedListMultimap<>();
-  private final Set<Pair<Long>> traceIdTimeStamps = new TreeSet<>(VALUE_2_DESCENDING);
-  private final Multimap<String, Pair<Long>> serviceToTraceIdTimeStamp =
-      new SortedByValue2Descending<>();
-  private final Multimap<String, String> serviceToSpanNames =
-      new LinkedHashSetMultimap<>();
+  /**
+   * Primary source of data is this map, which includes spans ordered descending by timestamp. All
+   * other maps are derived from the span values here.
+   *
+   * <p>In the future, we will bound data. This implies some mechanism of cascading deletes based on
+   * the pair used as the key here. One implementation could be to make one dataset have a strong
+   * reference to the pair of (traceId, timestamp), and others weak. For example, making this a
+   * weak reference and traceIdToTraceIdTimeStamps a strong one. In that case, deleting the trace ID
+   * from traceIdToTraceIdTimeStamps would lead to a purge of spans at GC time.
+   */
+  private final SortedMultimap<Pair<Long>, Span> spansByTraceIdTimeStamp =
+      new LinkedHashSetSortedMultimap<>(VALUE_2_DESCENDING);
+
+  /** This supports span lookup by {@link zipkin.Span#traceId lower 64-bits of the trace ID} */
+  private final SortedMultimap<Long, Pair<Long>> traceIdToTraceIdTimeStamps =
+      new LinkedHashSetSortedMultimap<>(Long::compareTo);
+  /**
+   * This supports span lookup by {@link zipkin.Endpoint#serviceName service name}.
+   *
+   * <p>QueryRequest.limit needs trace ids are returned in timestamp descending order.
+   */
+  private final SortedMultimap<String, Pair<Long>> serviceToTraceIdTimestamp =
+      new TreeSetSortedMultimap<>(String::compareTo, VALUE_2_DESCENDING);
```

Comment:
@joel-airspring FYI: I renamed this type to more generic as it keeps the comparators in one place.